src/pybind/mgr/cephadm/module.py

```diff
@@ -2980,7 +2980,10 @@ def upgrade_check(self, image: str, version: str) -> str:
         }
         for host, dm in self.cache.daemons.items():
             for name, dd in dm.items():
-                if image_info.image_id == dd.container_image_id:
+                if (
+                    (self.use_repo_digest and (any(d in (image_info.repo_digests or []) for d in (dd.container_image_digests or []))))
```

Comment:
can we break this condition in some logical/bolean checks? I find it hard to understand what are we actually checking (probabl a small comment could help :))

Reply:
I've moved some of the logic to class functions of the DaemonDescription class and added some comments. Hopefully that improves the clarity.

Reply:
thanks, I've just added a minor related comment :+1: